# Zomato Data Analysis 

Zomato, a popular Indian startup since 2008, is your go-to for exploring restaurants and ordering food online. Starting as a restaurant review hub, it now delivers a hassle-free dining experience. With a user-friendly setup, Zomato has become a global food discovery and delivery platform, making eating out or in a breeze.

In this initial phase, the focus is on cleansing the dataset to enhance its quality and reliability. Subsequently, forthcoming analyses will involve identifying outliers and conducting Exploratory Data Analysis (EDA) to glean further insights.

In [1]:
# Importing Major Libraries

import pandas as pd
import numpy as np 
from matplotlib import pyplot as plt 
import seaborn as sns
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Importing dataset 

df = pd.read_csv('zomato_ds1.csv')

In [3]:
# Viewing Dataset

df

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude  ... price_range  currency  \
0          Khandari  27.211450  78.002381  ...           2       Rs.   
1        Agra Cantt  27.160569  78.011583  ...           2       Rs.   
2          Shahganj  27.182938  77.979684  ...           1       Rs.   
3       Civil Lines  27.205668  78.004799  ...           1       Rs.   
4           Tajganj  27.157709  78.052421  ...           3       Rs.   
...             ...        ...        ...  ...         ...       ...   
211939    Fatehgunj  22.336931  73.192356  ...           2       Rs.   
211940   Karelibaug  22.322455  73.197203  ...           1       Rs.   
211941     Alkapuri  22.310563  73.171163  ...           2       Rs.   
211942        Akota  22.270027  73.143068  ...           2       Rs.   
211943     Vadiwadi  22.309935  73.158768  ...           2       Rs.   

                                               highlights aggregate_rating  \
0       ['Lunch', 'Takeaway Available', 'Credit Card',...              4.4   
1       ['Delivery', 'No Alcohol Available', 'Dinner',...              4.4   
2       ['No Alcohol Available', 'Dinner', 'Takeaway A...              4.2   
3       ['Takeaway Available', 'Credit Card', 'Lunch',...              4.3   
4       ['Lunch', 'Serves Alcohol', 'Cash', 'Credit Ca...              4.9   
...                                                   ...              ...   
211939  ['Dinner', 'Cash', 'Lunch', 'Delivery', 'Ind

- Rows are obscured by dots, indicating columns between them are present and not visible.

In [4]:
# Checking all columns

pd.set_option('display.max_columns',None)

In [5]:
# Rechecking entire dataset with all columns.

df

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude   zipcode  country_id  \
0          Khandari  27.211450  78.002381       NaN           1   
1        Agra Cantt  27.160569  78.011583  282001.0           1   
2          Shahganj  27.182938  77.979684  282010.0           1   
3       Civil Lines  27.205668  78.004799  282002.0           1   
4           Tajganj  27.157709  78.052421       NaN           1   
...             ...        ...        ...       ...         ...   
211939    Fatehgunj  22.336931  73.192356  390024.0           1   
211940   Karelibaug  22.322455  73.197203       NaN           1   
211941     Alkapuri  22.310563  73.171163       NaN           1   
211942        Akota  22.270027  73.143068       NaN           1   
211943     Vadiwadi  22.309935  73.158768  390007.0           1   

            locality_verbose  \
0             Khandari, Agra   
1           Agra Cantt, Agra   
2             Shahganj, Agra   
3          Civil Lines, Agra   
4              Tajganj, Agra   
...                      ...   
211939   Fatehgunj, Vadodara   
211940  Karelibaug, Vadodara   
211941    Alkapuri, Vadodara   
211942       Akota, Vadodara   
211943    Vadiwadi, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
1       North Indian, Mughlai, Rolls, Chinese, Fast Fo...   
2                                       Fast Food, Mithai   
3               Desserts, Bak

In [6]:
# Checking Information of Dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211944 entries, 0 to 211943
Data columns (total 26 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   res_id                211944 non-null  int64  
 1   name                  211944 non-null  object 
 2   establishment         211944 non-null  object 
 3   url                   211944 non-null  object 
 4   address               211810 non-null  object 
 5   city                  211944 non-null  object 
 6   city_id               211944 non-null  int64  
 7   locality              211944 non-null  object 
 8   latitude              211944 non-null  float64
 9   longitude             211944 non-null  float64
 10  zipcode               48757 non-null   object 
 11  country_id            211944 non-null  int64  
 12  locality_verbose      211944 non-null  object 
 13  cuisines              210553 non-null  object 
 14  timings               208070 non-null  object 
 15  

- We possess a total of **13 numerical columns**, with 4 designated as float type, incorporating decimal values, and the remaining 9 classified as integer type.
- We have **13 object-type columns**, possibly with text or categories. Further analysis will clarify their content.
- The dataset includes some blank columns.

In [7]:
# Checking description of dataset

df.describe()

res_id        city_id       latitude      longitude  country_id  \
count  2.119440e+05  211944.000000  211944.000000  211944.000000    211944.0   
mean   1.349411e+07    4746.785434      21.499758      77.615276         1.0   
std    7.883722e+06    5568.766386      22.781331       7.500104         0.0   
min    5.000000e+01       1.000000       0.000000       0.000000         1.0   
25%    3.301027e+06      11.000000      15.496071      74.877961         1.0   
50%    1.869573e+07      34.000000      22.514494      77.425971         1.0   
75%    1.881297e+07   11306.000000      26.841667      80.219323         1.0   
max    1.915979e+07   11354.000000   10000.000000      91.832769         1.0   

       average_cost_for_two    price_range  aggregate_rating          votes  \
count         211944.000000  211944.000000     211944.000000  211944.000000   
mean             595.812229       1.882535          3.395937     378.001864   
std              606.239363       0.892989          1.283642     925.333370   
min                0.000000       1.000000          0.000000     -18.000000   
25%              250.000000       1.000000          3.300000      16.000000   
50%              400.000000       2.000000          3.800000     100.000000   
75%              700.000000       2.000000          4.100000     362.000000   
max            30000.000000       4.000000          4.900000   42539.000000   

         photo_count  opentable_support       delivery  takeaway  
count  211944.000000           211896.0  211944.000000  211944.0  
mean      256.971224                0.0      -0.255907      -1.0  
std       867.668940                0.0       0.964172       0.0  
min         0.000000                0.0      -1.000000      -1.0  
25%         3.000000                0.0      -1.000000      -1.0  
50%        18.000000                0.0      -1.000000      -1.0  
75%       128.000000                0.0       1.000000      -1.0  
max     17702.000000                0.0       1.000000      -1.0

- The dataset comprises a total of 211,994 columns.
- We've examined the mean and standard deviation of the data. We'll explore their significance for potential future use.

### Checking for null values in the dataset.

In [8]:
df.isnull().sum()

res_id                       0
name                         0
establishment                0
url                          0
address                    134
city                         0
city_id                      0
locality                     0
latitude                     0
longitude                    0
zipcode                 163187
country_id                   0
locality_verbose             0
cuisines                  1391
timings                   3874
average_cost_for_two         0
price_range                  0
currency                     0
highlights                   0
aggregate_rating             0
rating_text                  0
votes                        0
photo_count                  0
opentable_support           48
delivery                     0
takeaway                     0
dtype: int64

- There are 134 inconsistencies in the columns of the address rows.
- There are 163,187 inconsistencies in the columns related to zip codes.
- There are 1,391 inconsistencies in the columns related to cuisines.
- There are 3,874 inconsistencies in the columns related to timings.
- There are 48 inconsistencies in the columns related to OpenTable support.

In [9]:
df.isnull().sum().sum()

168634

- There are a total of 168,634 inconsistencies in the data. It is essential to address and rectify these issues before proceeding with further analysis.

### Dealing with Null Values 

In [10]:
# Conducting a check for inconsistencies in the address column.

df[df.address.isnull()]

res_id                                     name  \
25475    2601516                                 CakeFizz   
78175   18313028  Pabrai's Fresh and Naturelle Ice Creams   
103427  18725005                test Moti Palace Township   
103433  18725012                           test Joshipura   
103575  18725005                test Moti Palace Township   
...          ...                                      ...   
197718  18041435                                 Lal Bagh   
197978  18041435                                 Lal Bagh   
198518  18041435                                 Lal Bagh   
198722  18041435                                 Lal Bagh   
199007  18041435                                 Lal Bagh   

              establishment  \
25475            ['Bakery']   
78175   ['Dessert Parlour']   
103427                   []   
103433                   []   
103575                   []   
...                     ...   
197718    ['Casual Dining']   
197978    ['Casual Dining']   
198518    ['Casual Dining']   
198722    ['Casual Dining']   
199007    ['Casual Dining']   

                                                      url address       city  \
25475   https://www.zomato.com/bhopal/cakefizz-jhangir...     NaN     Bhopal   
78175   https://www.zomato.com/hyderabad/pabrais-fresh...     NaN  Hyderabad   
103427  https://www.zomato.com/junagadh/junagadh/-moti...     NaN   Junagadh   
103433  https://www.zomato.com/junagadh/junagadh/-josh...     NaN   Junagadh   
103575  https://www.zomato.com/junagadh/junagadh/-moti...     NaN   Junagadh   
...                                                   ...     ...        ...   
197718  https://www.zomato.com/udaipur/lal-bagh-panchw...     NaN    Udaipur   
197978  https://www.zomato.com/udaipur/lal-bagh-panchw...     NaN    Udaipur   
198518  https://www.zomato.com/udaipur/lal-bagh-panchw...     NaN    Udaipur   
198722  https://www.zomato.com/udaipur/lal-bagh-panchw...     NaN    Udaipur   
199007  https://www.zomato.com/udaipur/lal-bagh-panchw...     NaN    Udaipur   

        city_id              locality   latitude  longitude   zipcode  \
25475        26           Jhangirabad  23.252214  77.435611       NaN   
78175         6              Madhapur  17.440883  78.393356  500081.0   
103427    11322  Moti Palace Township  21.498610  70.416507       NaN   
103433    11322             Joshipura  21.530117  70.453044       NaN   
103575    11322  Moti Palace Township  21.498610  70.416507       NaN   
...         ...                   ...        ...        ...       ...   
197718    11054             Panchwati  24.603519  73.687260       NaN   
197978    11054             Panchwati  24.603519  73.687260       NaN   
198518    11054             Panchwati  24.603519  73.687260       NaN   
198722    11054             Panchwati  24.603519  73.687260       NaN   
199007    11054             Panchwati  24.603519  73.687260       NaN   

        country_id                locality_verbose  \
25475            1             Jhangirabad, Bhopal   
78175            1             Madhapur, Hyderabad   
103427           1  Moti Palace Township, Junagadh   
103433           1             Joshipura, Junagadh   
103575           1  Moti Palace Township, Junagadh   
...            ...                             ...   
197718           1              Panchwati, Udaipur   
197978           1              Panchwati, Udaipur   
198518           1              Panchwati, Udaipur   
198722           1              Panchwati, Udaipur   
199007           1              Panchwati, Udaipur   

                                  cuisines                timings  \
25475                               Bakery               24 Hours   
78175                  Ice Cream, Desserts  11AM to 1AM (Mon-Sun)   
103427                                 NaN                    NaN   
103433                                 NaN                    NaN   
103575                                 NaN                    NaN   
...

In [11]:
# Comparing it with not null values 

df

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude   zipcode  country_id  \
0          Khandari  27.211450  78.002381       NaN           1   
1        Agra Cantt  27.160569  78.011583  282001.0           1   
2          Shahganj  27.182938  77.979684  282010.0           1   
3       Civil Lines  27.205668  78.004799  282002.0           1   
4           Tajganj  27.157709  78.052421       NaN           1   
...             ...        ...        ...       ...         ...   
211939    Fatehgunj  22.336931  73.192356  390024.0           1   
211940   Karelibaug  22.322455  73.197203       NaN           1   
211941     Alkapuri  22.310563  73.171163       NaN           1   
211942        Akota  22.270027  73.143068       NaN           1   
211943     Vadiwadi  22.309935  73.158768  390007.0           1   

            locality_verbose  \
0             Khandari, Agra   
1           Agra Cantt, Agra   
2             Shahganj, Agra   
3          Civil Lines, Agra   
4              Tajganj, Agra   
...                      ...   
211939   Fatehgunj, Vadodara   
211940  Karelibaug, Vadodara   
211941    Alkapuri, Vadodara   
211942       Akota, Vadodara   
211943    Vadiwadi, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
1       North Indian, Mughlai, Rolls, Chinese, Fast Fo...   
2                                       Fast Food, Mithai   
3               Desserts, Bak

- It appears that there are no discernible patterns or insights that would facilitate a meaningful replacement for the inconsistencies in the address column. Considering the primary focus of the address column on the actual restaurant address, alternatives such as using city or locality may not be suitable.

Certainly, **using "Unknown" as a placeholder for the inconsistencies in the address column** is a pragmatic approach. This temporary solution allows for continued analysis and can be revisited as needed.

In [12]:
# Filling NUll values in Address with placeholder "Unknown"

df.address = df.address.fillna('Unknown')

In [13]:
df.address.isnull().sum()

0

- **Address columns** are now void of null entries, ensuring a comprehensive and accurate dataset.

In [14]:
# Conducting a check for inconsistencies in the zipcodes.

df[df.zipcode.isnull()]

res_id                         name      establishment  \
0        3400299                  Bikanervala    ['Quick Bites']   
4        3401744  The Salt Cafe Kitchen & Bar  ['Casual Dining']   
5        3400275               Domino's Pizza    ['Quick Bites']   
7        3400368               Domino's Pizza    ['Quick Bites']   
8        3401284                   Cake House         ['Bakery']   
...          ...                          ...                ...   
211937  18855810          Biryani aur Baatein  ['Casual Dining']   
211938  18662583                  Wok On Fire  ['Casual Dining']   
211940   3200996                   Raju Omlet    ['Quick Bites']   
211941  18984164             The Grand Thakar  ['Casual Dining']   
211942   3201138                       Subway    ['Quick Bites']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
5       https://www.zomato.com/agra/dominos-pizza-civi...   
7       https://www.zomato.com/agra/dominos-pizza-sika...   
8       https://www.zomato.com/agra/cake-house-2-civil...   
...                                                   ...   
211937  https://www.zomato.com/vadodara/biryani-aur-ba...   
211938  https://www.zomato.com/vadodara/wok-on-fire-fa...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
5       114/23 G, Deep Shikha Complex, Sanjay Place, C...      Agra       34   
7                   Plot C-1/6, Sector 13, Sikandra, Agra      Agra       34   
8       23/301, Wazirpura Rd, Judge Compound Chowraha,...      Agra       34   
...                                                   ...       ...      ...   
211937  Shop 14, Atlantis K-10, A Wing, Genda Circle R...  Vadodara       32   
211938  Ground Floor 1, Rossette Building, Opposite Se...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   

           locality   latitude  longitude zipcode  country_id  \
0          Khandari  27.211450  78.002381     NaN           1   
4           Tajganj  27.157709  78.052421     NaN           1   
5       Civil Lines  27.201516  78.007556     NaN           1   
7          Sikandra  27.203930  77.954260     NaN           1   
8       Civil Lines  27.204148  78.009025     NaN           1   
...             ...        ...        ...     ...         ...   
211937     Alkapuri  22.317746  73.168043     NaN           1   
211938    Fatehgunj  22.323357  73.187461     NaN           1   
211940   Karelibaug  22.322455  73.197203     NaN           1   
211941     Alkapuri  22.310563  73.171163     NaN           1   
211942        Akota  22.270027  73.143068     NaN           1   

            locality_verbose  \
0             Khandari, Agra   
4              Tajganj, Agra   
5          Civil Lines, Agra   
7             Sikandra, Agra   
8          Civil Lines, Agra   
...                      ...   
211937    Alkapuri, Vadodara   
211938   Fatehgunj, Vadodara   
211940  Karelibaug, Vadodara   
211941    Alkapuri, Vadodara   
211942       Akota, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
4                      North Indian, Continental, Italian   
5                                        Pizza, Fast Food   
7                                        Pizza, Fast Food   
8                           

In [15]:
# Comparing it with the dataset 

df

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude   zipcode  country_id  \
0          Khandari  27.211450  78.002381       NaN           1   
1        Agra Cantt  27.160569  78.011583  282001.0           1   
2          Shahganj  27.182938  77.979684  282010.0           1   
3       Civil Lines  27.205668  78.004799  282002.0           1   
4           Tajganj  27.157709  78.052421       NaN           1   
...             ...        ...        ...       ...         ...   
211939    Fatehgunj  22.336931  73.192356  390024.0           1   
211940   Karelibaug  22.322455  73.197203       NaN           1   
211941     Alkapuri  22.310563  73.171163       NaN           1   
211942        Akota  22.270027  73.143068       NaN           1   
211943     Vadiwadi  22.309935  73.158768  390007.0           1   

            locality_verbose  \
0             Khandari, Agra   
1           Agra Cantt, Agra   
2             Shahganj, Agra   
3          Civil Lines, Agra   
4              Tajganj, Agra   
...                      ...   
211939   Fatehgunj, Vadodara   
211940  Karelibaug, Vadodara   
211941    Alkapuri, Vadodara   
211942       Akota, Vadodara   
211943    Vadiwadi, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
1       North Indian, Mughlai, Rolls, Chinese, Fast Fo...   
2                                       Fast Food, Mithai   
3               Desserts, Bak

In [16]:
# Trying to seek any relationship between zipcode and locality 

selected_columns = ['zipcode', 'locality']
df_filtered = df[selected_columns]

In [17]:
df_filtered

zipcode     locality
0            NaN     Khandari
1       282001.0   Agra Cantt
2       282010.0     Shahganj
3       282002.0  Civil Lines
4            NaN      Tajganj
...          ...          ...
211939  390024.0    Fatehgunj
211940       NaN   Karelibaug
211941       NaN     Alkapuri
211942       NaN        Akota
211943  390007.0     Vadiwadi

[211944 rows x 2 columns]

In [18]:
df_filtered[df_filtered.locality=="Khandari"]

zipcode  locality
0          NaN  Khandari
33         NaN  Khandari
65    282002.0  Khandari
68         NaN  Khandari
72    282003.0  Khandari
...        ...       ...
2556       NaN  Khandari
2569  282003.0  Khandari
2575  282002.0  Khandari
2579       NaN  Khandari
2581       NaN  Khandari

[156 rows x 2 columns]

In some locations, zipcodes are listed along with their corresponding localities, while in other cases, those same localities have empty or blank zipcode columns.
- For instance, in the "Khandari" locality, we observe both null values and assigned values in the zipcode column.

We aim to determine the most frequent (mode) zipcodes associated with non-null values in the locality column, then use this mode to replace the null values in the corresponding zipcode column.

In [19]:
# Creating a function to replace null values in the "zipcode" column.

def fill_missing_zipcodes(dataframe):
# Create a dictionary to store locality-zipcode mappings
    locality_zipcode_mapping = {}

# Iterate through the rows to populate the dictionary
    for index, row in dataframe.iterrows():
        locality = row['locality']
        zipcode = row['zipcode']

# Update the mapping only if the zipcode is not null
        if not pd.isnull(zipcode):
            locality_zipcode_mapping[locality] = zipcode

# Iterate again to fill missing zipcodes
    for index, row in dataframe.iterrows():
        locality = row['locality']
        zipcode = row['zipcode']

# Check if the zipcode is null and if the locality has a mapping
        if pd.isnull(zipcode) and locality in locality_zipcode_mapping:
            dataframe.at[index, 'zipcode'] = locality_zipcode_mapping[locality]

# Replace missing zipcodes with 0 where there is no mapping
    dataframe['zipcode'].fillna(0, inplace=True)

In [20]:
# Running Function to remove null values from dataset

fill_missing_zipcodes(df)

In [21]:
# Checking if any null values are left in zipcode

df.zipcode.isnull().sum()

0

- No null values remain in the "zipcode" column.

In [22]:
# Handling missing values in cuisine entries

df[df.cuisines.isnull()].head()

res_id                        name      establishment  \
1140    3401673  Sri Dauji Mishthan Bhandar     ['Sweet Shop']   
7770   18689668          Vedanta Restaurant  ['Casual Dining']   
8000   18689668          Vedanta Restaurant  ['Casual Dining']   
11389  19008079                Shree sweets                 []   
11390  19008131     Aditya Mishthan Bhandar                 []   

                                                     url  \
1140   https://www.zomato.com/agra/sri-dauji-mishthan...   
7770   https://www.zomato.com/ajmer/vedanta-restauran...   
8000   https://www.zomato.com/ajmer/vedanta-restauran...   
11389  https://www.zomato.com/allahabad/shree-sweets-...   
11390  https://www.zomato.com/allahabad/aditya-mishth...   

                                                 address       city  city_id  \
1140   F11, Near Agrawal Hospital, Langre Ki Chowki, ...       Agra       34   
7770   Hotel Metro Inn, Near Saint Francis Hospital, ...      Ajmer    11303   
8000   Hotel Metro Inn, Near Saint Francis Hospital, ...      Ajmer    11303   
11389               6, Rambagh Hanuman Mandir, Allahabad  Allahabad       24   
11390   153, South Malaka, Allahabad, Rambagh, Allahabad  Allahabad       24   

                locality   latitude  longitude     zipcode  country_id  \
1140         Kamla Nagar  27.207933  78.025556           0           1   
7770   Railway Quarters   26.448262  74.638991  Apna  baza           1   
8000   Railway Quarters   26.448262  74.638991  Apna  baza           1   
11389            Rambagh  25.439095  81.847112    211003.0           1   
11390            Rambagh  25.440297  81.844704    211003.0           1   

               locality_verbose cuisines         timings  \
1140          Kamla Nagar, Agra      NaN   7 AM to 11 PM   
7770   Railway Quarters , Ajmer      NaN  10 AM to 11 PM   
8000   Railway Quarters , Ajmer      NaN  10 AM to 11 PM   
11389        Rambagh, Allahabad      NaN   7 AM to 10 PM   
11390        Rambagh, Allahabad      NaN   7 AM to 11 PM   

       average_cost_for_two  price_range currency  \
1140                    200            1      Rs.   
7770                      0            1      Rs.   
8000                      0            1      Rs.   
11389                   100            1      Rs.   
11390                   100            1      Rs.   

                                              highlights  aggregate_rating  \
1140   ['Breakfast', 'Cash', 'Takeaway Available', 'I...               3.6   
7770   ['Cash', 'Debit Card', 'Dinner', 'Takeaway Ava...               0.0   
8000   ['Cash', 'Debit Card', 'Dinner', 'Takeaway Ava...               0.0   
11389                                                 []               0.0   
11390                                                 []               0.0   

      rating_text  votes  photo_count  opentable_support  delivery  takeaway  
1140         Good     23            0                0.0        -1        -1  
7770    Not rated      3            3                0.0        -1        -1  
8000    Not rated      3            3                0.0        -1        -1  
11389   Not rated      1            0                0.0        -1        -1  
11390   Not rated      1            0                0.0        -1        -1

In [23]:
# Comparing rows with null values to rows where cuisines are not null.

df

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude   zipcode  country_id  \
0          Khandari  27.211450  78.002381  282002.0           1   
1        Agra Cantt  27.160569  78.011583  282001.0           1   
2          Shahganj  27.182938  77.979684  282010.0           1   
3       Civil Lines  27.205668  78.004799  282002.0           1   
4           Tajganj  27.157709  78.052421  282001.0           1   
...             ...        ...        ...       ...         ...   
211939    Fatehgunj  22.336931  73.192356  390024.0           1   
211940   Karelibaug  22.322455  73.197203  390018.0           1   
211941     Alkapuri  22.310563  73.171163  390007.0           1   
211942        Akota  22.270027  73.143068  390020.0           1   
211943     Vadiwadi  22.309935  73.158768  390007.0           1   

            locality_verbose  \
0             Khandari, Agra   
1           Agra Cantt, Agra   
2             Shahganj, Agra   
3          Civil Lines, Agra   
4              Tajganj, Agra   
...                      ...   
211939   Fatehgunj, Vadodara   
211940  Karelibaug, Vadodara   
211941    Alkapuri, Vadodara   
211942       Akota, Vadodara   
211943    Vadiwadi, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
1       North Indian, Mughlai, Rolls, Chinese, Fast Fo...   
2                                       Fast Food, Mithai   
3               Desserts, Bak

 In the absence of discernible patterns, replacing null values with an "No Description" placeholder.

In [24]:
df.cuisines = df.cuisines.fillna('No Description')

In [25]:
# Checking for Null values in cuisines 

df.cuisines.isnull().sum()

0

- No null values remain in the "cuisine" column.

In [26]:
# Handling missing values in timing entries.

df[df.timings.isnull()]

res_id                             name      establishment  \
126      3400957                            Anise    ['Fine Dining']   
214     19032611    Om Sai Lassi And Juice Corner  ['Beverage Shop']   
220     19050396                    Agarwal Lassi          ['Shack']   
297     19031666                    chai ki tapri           ['Café']   
383      3400322    Filmy Zayka - Hotel Alleviate  ['Casual Dining']   
...          ...                              ...                ...   
207778  19065540              Bake N Break Bakery         ['Bakery']   
208022  19065540              Bake N Break Bakery         ['Bakery']   
208465  19065540              Bake N Break Bakery         ['Bakery']   
209415   3200082  Salsa Habanero Mexican Taqueria  ['Casual Dining']   
210951  18732612       Elysian: The Bistro Lounge  ['Casual Dining']   

                                                      url  \
126     https://www.zomato.com/agra/anise-tajganj?utm_...   
214     https://www.zomato.com/agra/om-sai-lassi-and-j...   
220     https://www.zomato.com/agra/agarwal-lassi-civi...   
297     https://www.zomato.com/agra/chai-ki-tapri-sika...   
383     https://www.zomato.com/agra/filmy-zayka-hotel-...   
...                                                   ...   
207778  https://www.zomato.com/visakhapatnam/bake-n-br...   
208022  https://www.zomato.com/visakhapatnam/bake-n-br...   
208465  https://www.zomato.com/visakhapatnam/bake-n-br...   
209415  https://www.zomato.com/vadodara/salsa-habanero...   
210951  https://www.zomato.com/elysiantbl?utm_source=a...   

                                                  address      city  city_id  \
126     Courtyard by Marriott Agra, Taj Nagri , Phase ...      Agra       34   
214                        Bodla Chauraya, Shahganj, Agra      Agra       34   
220     22/156chandi wali gali kothi near paliwal park...      Agra       34   
297     shop no 2 hno 216 sec- 6a avas vikas colony si...      Agra       34   
383     6/27, Yamuna Kinara Road, Belanganj, Civil Lin...      Agra       34   
...                                                   ...       ...      ...   
207778  Door 21-2-13/14, Sector 9, MVP Colony, Visakha...     Vizag       28   
208022  Door 21-2-13/14, Sector 9, MVP Colony, Visakha...     Vizag       28   
208465  Door 21-2-13/14, Sector 9, MVP Colony, Visakha...     Vizag       28   
209415  144/145 Villa D Este Next to Madhav Tower, Sam...  Vadodara       32   
210951  FF 108 Golden Icon, Opposite BSNL, Chakli Circ...  Vadodara       32   

           locality   latitude  longitude   zipcode  country_id  \
126         Tajganj  27.156386  78.059525  282001.0           1   
214        Shahganj  27.188775  77.962173  282010.0           1   
220     Civil Lines  27.199784  78.017700    492006           1   
297        Sikandra  27.199757  77.958770  282007.0           1   
383     Civil Lines  27.189864  78.024778    492006           1   
...             ...        ...        ...       ...         ...   
207778   MVP Colony  17.716327  83.311777  530017.0           1   
208022   MVP Colony  17.716327  83.311777  530017.0           1   
208465   MVP Colony  17.716327  83.311777  530017.0           1   
209415   Subhanpura  22.318358  73.151159  390023.0           1   
210951     Vadiwadi  22.307717  73.165685  390007.0           1   

            locality_verbose  \
126            Tajganj, Agra   
214           Shahganj, Agra   
220        Civil Lines, Agra   
297           Sikandra, Agra   
383        Civil Lines, Agra   
...                      ...   
207778     MVP Colony, Vizag   
208022     MVP Colony, Vizag   
208465     MVP Colony, Vizag   
209415  Subhanpura, Vadodara   
210951    Vadiwadi, Vadodara   

                                                 cuisines timings  \
126                                          North Indian     NaN   
214                                                Juices     NaN   
220                                           Street F

In [27]:
# Comparing it with Not NUll columns

df

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude   zipcode  country_id  \
0          Khandari  27.211450  78.002381  282002.0           1   
1        Agra Cantt  27.160569  78.011583  282001.0           1   
2          Shahganj  27.182938  77.979684  282010.0           1   
3       Civil Lines  27.205668  78.004799  282002.0           1   
4           Tajganj  27.157709  78.052421  282001.0           1   
...             ...        ...        ...       ...         ...   
211939    Fatehgunj  22.336931  73.192356  390024.0           1   
211940   Karelibaug  22.322455  73.197203  390018.0           1   
211941     Alkapuri  22.310563  73.171163  390007.0           1   
211942        Akota  22.270027  73.143068  390020.0           1   
211943     Vadiwadi  22.309935  73.158768  390007.0           1   

            locality_verbose  \
0             Khandari, Agra   
1           Agra Cantt, Agra   
2             Shahganj, Agra   
3          Civil Lines, Agra   
4              Tajganj, Agra   
...                      ...   
211939   Fatehgunj, Vadodara   
211940  Karelibaug, Vadodara   
211941    Alkapuri, Vadodara   
211942       Akota, Vadodara   
211943    Vadiwadi, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
1       North Indian, Mughlai, Rolls, Chinese, Fast Fo...   
2                                       Fast Food, Mithai   
3               Desserts, Bak

- The 'timing' column is of object type, encompassing string values that include both timings and days of the week when the establishment is open.

In [28]:
df.timings = df.timings.fillna("No Info")

In [29]:
df.timings.isnull().sum()

0

The **timings column** is now free of any null values.

In [30]:
# Handling null values in opentable_support entries

df.opentable_support

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
         ... 
211939    0.0
211940    0.0
211941    0.0
211942    0.0
211943    0.0
Name: opentable_support, Length: 211944, dtype: float64

In [31]:
df.opentable_support.mean()

0.0

Given that the 'opentable_support' column has all zero values, as evident from their mean being zero, we are filling null values with zero for consistency.

In [32]:
df.opentable_support = df.opentable_support.fillna(0)

In [33]:
df.opentable_support.isnull().sum()

0

- No null values remain in the "opentable_support" column.

### Verifying the entire dataset for any remaining inconsistencies.

In [34]:
df.isnull().sum()

res_id                  0
name                    0
establishment           0
url                     0
address                 0
city                    0
city_id                 0
locality                0
latitude                0
longitude               0
zipcode                 0
country_id              0
locality_verbose        0
cuisines                0
timings                 0
average_cost_for_two    0
price_range             0
currency                0
highlights              0
aggregate_rating        0
rating_text             0
votes                   0
photo_count             0
opentable_support       0
delivery                0
takeaway                0
dtype: int64

The entire dataset is now void of any null values.

#### Removing unnecessary columns from the dataset for improved clarity.

In [35]:
df

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude   zipcode  country_id  \
0          Khandari  27.211450  78.002381  282002.0           1   
1        Agra Cantt  27.160569  78.011583  282001.0           1   
2          Shahganj  27.182938  77.979684  282010.0           1   
3       Civil Lines  27.205668  78.004799  282002.0           1   
4           Tajganj  27.157709  78.052421  282001.0           1   
...             ...        ...        ...       ...         ...   
211939    Fatehgunj  22.336931  73.192356  390024.0           1   
211940   Karelibaug  22.322455  73.197203  390018.0           1   
211941     Alkapuri  22.310563  73.171163  390007.0           1   
211942        Akota  22.270027  73.143068  390020.0           1   
211943     Vadiwadi  22.309935  73.158768  390007.0           1   

            locality_verbose  \
0             Khandari, Agra   
1           Agra Cantt, Agra   
2             Shahganj, Agra   
3          Civil Lines, Agra   
4              Tajganj, Agra   
...                      ...   
211939   Fatehgunj, Vadodara   
211940  Karelibaug, Vadodara   
211941    Alkapuri, Vadodara   
211942       Akota, Vadodara   
211943    Vadiwadi, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
1       North Indian, Mughlai, Rolls, Chinese, Fast Fo...   
2                                       Fast Food, Mithai   
3               Desserts, Bak

In [36]:
df.country_id.nunique()

1

- Since all the data is from the same country, the "country_id" column is unnecessary and will be dropped for simplification.

In [37]:
df.res_id.nunique()

55568

- Observing significant duplicacy in the "res_id" column, resembling a bubble for a region. Given this, the "res_id" will be disregarded, and index numbers will be used for clarity.

In [38]:
df.currency.nunique()

1

- Given that the country is consistent for all rows and the restaurants only accept Rs./INR, the "currency" column is redundant and will be dropped for simplicity.

In [39]:
df.opentable_support.nunique()

1

- As the "opentable_support" column consistently holds zero values across all rows, it does not provide valuable insights. Consequently, this column will be dropped for efficiency.

In [40]:
# Dropping unnecessary rows from dataset 

df = df.drop(["res_id","country_id","currency","opentable_support"],axis=1)

In [41]:
df

name      establishment  \
0                          Bikanervala    ['Quick Bites']   
1       Mama Chicken Mama Franky House    ['Quick Bites']   
2                        Bhagat Halwai    ['Quick Bites']   
3                        Bhagat Halwai    ['Quick Bites']   
4          The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...                                ...                ...   
211939  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940                      Raju Omlet    ['Quick Bites']   
211941                The Grand Thakar  ['Casual Dining']   
211942                          Subway    ['Quick Bites']   
211943     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude   zipcode      locality_verbose  \
0          Khandari  27.211450  78.002381  282002.0        Khandari, Agra   
1        Agra Cantt  27.160569  78.011583  282001.0      Agra Cantt, Agra   
2          Shahganj  27.182938  77.979684  282010.0        Shahganj, Agra   
3       Civil Lines  27.205668  78.004799  282002.0     Civil Lines, Agra   
4           Tajganj  27.157709  78.052421  282001.0         Tajganj, Agra   
...             ...        ...        ...       ...                   ...   
211939    Fatehgunj  22.336931  73.192356  390024.0   Fatehgunj, Vadodara   
211940   Karelibaug  22.322455  73.197203  390018.0  Karelibaug, Vadodara   
211941     Alkapuri  22.310563  73.171163  390007.0    Alkapuri, Vadodara   
211942        Akota  22.270027  73.143068  390020.0       Akota, Vadodara   
211943     Vadiwadi  22.309935  73.158768  390007.0    Vadiwadi, Vadodara   

                                                 cuisines  \
0       North Indian, South Indian, Mithai, Street Foo...   
1       North Indian, Mughlai, Rolls, Chinese, Fast Fo...   
2                                       Fast Food, Mithai   
3               Desserts, Bakery, Fast Food, South Indian   
4                      North Indian, Continental, Italian   
...                                                   ...   
211939                                       North Indian   
211940                                          Fast Food   
211941                    Gujarati, North Indian, Chinese   
211942                         Fast Food, Sandwich, Salad   
211943     

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211944 entries, 0 to 211943
Data columns (total 22 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   name                  211944 non-null  object 
 1   establishment         211944 non-null  object 
 2   url                   211944 non-null  object 
 3   address               211944 non-null  object 
 4   city                  211944 non-null  object 
 5   city_id               211944 non-null  int64  
 6   locality              211944 non-null  object 
 7   latitude              211944 non-null  float64
 8   longitude             211944 non-null  float64
 9   zipcode               211944 non-null  object 
 10  locality_verbose      211944 non-null  object 
 11  cuisines              211944 non-null  object 
 12  timings               211944 non-null  object 
 13  average_cost_for_two  211944 non-null  int64  
 14  price_range           211944 non-null  int64  
 15  

Currently, the dataset comprises a total of 22 columns, including 10 numerical columns and 12 object-type columns.

- Noticing that the "zipcode" column is turned object type after using **fill_missing_zipcodes() function** , it will be converted to int type for accuracy.

In [43]:
df['zipcode'] = df['zipcode'].astype(int)

ValueError: invalid literal for int() with base 10: '282002.0'

- The error suggests that there might be non-integer values (e.g., NaN or string values) in the "zipcode" column that cannot be directly converted to integers. 
- There might be some error with code of block

In [ ]:
# Trying to identify the problematic values by attempting to convert each value individually using a try-except block.

In [ ]:
for value in df['zipcode']:
    try:
        int(value)
    except ValueError:
        print(f"Cannot convert {value} to int")
        
# /////TURNED OFF THIS COMMAND FOR EASE IN READING //////

- "Apna  baza, absoluteba, Bangalore, Shop no. S" are few phrases we encountered manually 

In [44]:
# Convert the "zipcode" column to strings so that we can use isalpha()

df['zipcode'] = df['zipcode'].astype(str)

In [45]:
# Custom function to check if the input contains only alphabets or words
def contains_only_alphabets(value):
    return value.isalpha()

# Apply the function to filter rows
alphabetic_rows = df[df['zipcode'].apply(contains_only_alphabets)]

# Display the rows where the "zipcode" column contains only alphabets or words
alphabetic_rows

name        establishment  \
19951   AB's - Absolute Barbecues    ['Casual Dining']   
20497                    Cake Box           ['Bakery']   
20614            Lakeview Milkbar  ['Dessert Parlour']   
21490                  Tiffin Box             ['Mess']   
21515   Sri Ganesh Juice Junction    ['Beverage Shop']   
21549   Sri Ganesh Juice Junction    ['Beverage Shop']   
23155   AB's - Absolute Barbecues    ['Casual Dining']   
23209   AB's - Absolute Barbecues    ['Casual Dining']   
24583  Blue Tokai Coffee Roasters             ['Café']   

                                                     url  \
19951  https://www.zomato.com/bangalore/abs-absolute-...   
20497  https://www.zomato.com/bangalore/cake-box-1-ko...   
20614  https://www.zomato.com/bangalore/lakeview-milk...   
21490  https://www.zomato.com/bangalore/tiffin-box-ko...   
21515  https://www.zomato.com/bangalore/sri-ganesh-ju...   
21549  https://www.zomato.com/bangalore/sri-ganesh-ju...   
23155  https://www.zomato.com/bangalore/abs-absolute-...   
23209  https://www.zomato.com/bangalore/abs-absolute-...   
24583  https://www.zomato.com/bangalore/blue-tokai-co...   

                                                 address       city  city_id  \
19951  1st Floor, GRS Towers, Above Spencers Hyper Ma...  Bangalore        4   
20497  74/2A, K.R. Garden, Koramangala 8th Block, Ban...  Bangalore        4   
20614  8th Block, Opposite Bethany School, 80 Feet Ro...  Bangalore        4   
21490  214, 3rd Main Road, Koramangala 8th Block, Ban...  Bangalore        4   
21515  537, 1st Main, Near Bethani School, 8th Block,...  Bangalore        4   
21549  513, Opposite Ganesh Temple, Koramangala 8th B...  Bangalore        4   
23155  1st Floor, GRS Towers, Above Spencers Hyper Ma...  Bangalore        4   
23209  1st Floor, GRS Towers, Above Spencers Hyper Ma...  Bangalore        4   
24583  583, 80 Feet Road, Opposite Bethany High Schoo...  Bangalore        4   

                    locality   latitude  longitude     zipcode  \
19951          Sarjapur Road  12.918717  77.669925  absoluteba   
20497  Koramangala 8th Block  12.943477  77.620535   Bangalore   
20614  Koramangala 8th Block  12.940799  77.620468   Bangalore   
21490  Koramangala 8th Block  12.941822  77.616169   Bangalore   
21515  Koramangala 8th Block  12.940879  77.619751   Bangalore   
21549  Koramangala 8th Block  12.941632  77.621547   Bangalore   
23155          Sarjapur Road  12.918717  77.669925  absoluteba   
23209          Sarjapur Road  12.918717  77.669925  absoluteba   
24583  Koramangala 8th Block  12.940603  77.620291   Bangalore   

                       locality_verbose  \
19951          Sarjapur Road, Bangalore   
20497  Koramangala 8th Block, Bangalore   
20614  Koramangala 8th Block, Bangalore   
21490  Koramangala 8th Block, Bangalore   
21515  Koramangala 8th Block, Bangalore   
21549  Koramangala 8th Block, Bangalore   
23155          Sarjapur Road, Bangalore   
23209          Sarjapur Road, Bangalore   
24583  Koramangala 8th Block, Bangalore   

                                         cuisines  \
19951  European, Mediterranean, North Indian, BBQ   
20497                            Bakery, Desserts   
20614                              Desserts, Cafe   
21490                     South Indian, Beverages   
21515                Beverages, Juices, Fast Food   
21549                           Juices, Fast Food   
23155  European, Mediterranean, North Indian, BBQ   
23209  European, Mediterranean, North Indian, BBQ   
24583                                Coffee, Cafe   

                                       timings  average_cost_for_two  \
19951  12 Noon to 4 PM, 6:30 PM to 12 Midnight                  1600   
20497                         10:30 AM to 2 AM                   400   
20614                   9 AM to 1 AM (Mon-Sun)                   500   
21490                             8 AM to 9 PM                   400   
21515                     7am – 11pm (Mon-Sun)                 

- Now we are able to see string data which is spoiling the column 

In [46]:
# Dealing with non_numerical columns in this and replacing non-numerical columns with 0

# Making a variable and identifying rows where the "zipcode" column contains only alphabets or words
non_numeric_rows = df['zipcode'].apply(contains_only_alphabets)

# Replace non-numeric values with zero
df.loc[non_numeric_rows, 'zipcode'] = '0'

In [47]:
# Checking if there any non_numerical values are left in the data 

alphabetic_rows = df[df['zipcode'].apply(contains_only_alphabets)]

alphabetic_rows

Empty DataFrame
Columns: [name, establishment, url, address, city, city_id, locality, latitude, longitude, zipcode, locality_verbose, cuisines, timings, average_cost_for_two, price_range, highlights, aggregate_rating, rating_text, votes, photo_count, delivery, takeaway]
Index: []

- Now we have all numerical columns 

In [48]:
# changing dtype from str to int for better clarity 

df['zipcode'] = df['zipcode'].astype(int)

ValueError: invalid literal for int() with base 10: '282002.0'

- We have removed all the alphanumeric rows but still we are getting an error
- Let us first change it into float for better understanding

In [49]:
# Convert the "zipcode" column to float
df['zipcode'] = df['zipcode'].astype(float)

ValueError: could not convert string to float: '380 006'

- As we can see few data contains spaces as well. It is taking it as String 

In [50]:
# Replace spaces in the "zipcode" column
df['zipcode'] = df['zipcode'].str.replace(' ', '')

In [51]:
# Convert the "zipcode" column to float
df['zipcode'] = df['zipcode'].astype(float)

ValueError: could not convert string to float: 'Apnabaza'

- We can see an str Apnabaza is left during process, we need to replace it with zero 

In [52]:
# Convert the "zipcode" column to numeric, coercing errors to NaN
df['zipcode'] = pd.to_numeric(df['zipcode'], errors='coerce')

In [53]:
# Replace NaN values with 0
df['zipcode'].fillna(0, inplace=True)

In [54]:
# Convert the "zipcode" column to integers
df['zipcode'] = df['zipcode'].astype(int)

In [55]:
df.zipcode.dtype

dtype('int32')

**The "zipcode" column's data type has been formally and efficiently updated from object to integer.**

In [56]:
# Verifying the count of rows converted to zero during the process.

turned_zero = df.zipcode==0

In [57]:
turned_zero.value_counts()

zipcode
False    166806
True      45138
Name: count, dtype: int64

### NOTE:

A total of 45,138 rows have been converted to zero during the process. However, **it's noteworthy that 118,049 rows have been successfully retained.** Future considerations will be given to how to address or handle the rows that have been converted to zero.

### Identifying and addressing duplicates in the dataset.

In [58]:
df[df.duplicated]

name      establishment  \
101              Peshawri - ITC Mughal    ['Fine Dining']   
116              Taj Bano - ITC Mughal    ['Fine Dining']   
140                     Pinch Of Spice  ['Casual Dining']   
141                     Pinch Of Spice  ['Casual Dining']   
142                         Urban Deck  ['Casual Dining']   
...                                ...                ...   
211937             Biryani aur Baatein  ['Casual Dining']   
211938                     Wok On Fire  ['Casual Dining']   
211939  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211941                The Grand Thakar  ['Casual Dining']   
211943     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
101     https://www.zomato.com/agra/peshawri-itc-mugha...   
116     https://www.zomato.com/agra/taj-bano-itc-mugha...   
140     https://www.zomato.com/agra/pinch-of-spice-civ...   
141     https://www.zomato.com/agra/pinch-of-spice-taj...   
142     https://www.zomato.com/agra/urban-deck-2-civil...   
...                                                   ...   
211937  https://www.zomato.com/vadodara/biryani-aur-ba...   
211938  https://www.zomato.com/vadodara/wok-on-fire-fa...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
101             ITC Mughal, Fatehabad Road, Tajganj, Agra      Agra       34   
116             ITC Mughal, Fatehabad Road, Tajganj, Agra      Agra       34   
140     23/453, Opposite Sanjay Cinema, Wazipura Road,...      Agra       34   
141                 1076/2, Fatehabad Road, Tajganj, Agra      Agra       34   
142     5th Floor, The P L Palace Hotel, MG Road, Sanj...      Agra       34   
...                                                   ...       ...      ...   
211937  Shop 14, Atlantis K-10, A Wing, Genda Circle R...  Vadodara       32   
211938  Ground Floor 1, Rossette Building, Opposite Se...  Vadodara       32   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

                   locality   latitude  longitude  zipcode  \
101     ITC Mughal, Tajganj  27.161150  78.043993        0   
116     ITC Mughal, Tajganj  27.161132  78.044022        0   
140             Civil Lines  27.201735  78.007625   492006   
141                 Tajganj  27.159649  78.043304   282010   
142             Civil Lines  27.199573  78.003699   282005   
...                     ...        ...        ...      ...   
211937             Alkapuri  22.317746  73.168043   390007   
211938            Fatehgunj  22.323357  73.187461   390024   
211939            Fatehgunj  22.336931  73.192356   390024   
211941             Alkapuri  22.310563  73.171163   390007   
211943             Vadiwadi  22.309935  73.158768   390007   

                 locality_verbose                            cuisines  \
101     ITC Mughal, Tajganj, Agra               North Indian, Mughlai   
116     ITC Mughal, Tajganj, Agra               Mughlai, North Indian   
140             Civil Lines, Agra      North Indian, Chinese, Mughlai   
141                 Tajganj, Agra  North Indian, Chinese, Continental   
142             Civil Lines, Agra      Mexican, Italian, North Indian   
...                           ...                                 ...   
211937         Alkapuri, Vadodara           Biryani, Mughlai, Chinese   
211938        Fatehgunj, Vadodara                      Chinese, Asian   
211939        Fatehgunj, Vadodara                        North Indian   
211941         Alkapuri, Vadodara     Gujarati, North Indian, Chinese   
211943         Vadiwadi, Vadodara          Cafe, Healthy Fo

- With a dataset comprising 151,533 rows, we have observed that a considerable amount of data appears to be legitimate. While acknowledging the potential presence of exactly identical rows, we have decided not to remove duplicates, recognizing that eliminating such a significant amount of data could adversely impact the dataset's integrity and compromise future analyses.

#### The data has been successfully cleaned, devoid of inconsistencies. We will proceed to save it as a new CSV file.

In [59]:
# Saving the refined dataset as a new .csv file for subsequent analysis and error identification.

df.to_csv('Zomato_CleanData.csv')

# SUMMARY:

Throughout the data cleaning process, several key steps were undertaken to enhance the quality and integrity of the dataset. The initial assessment involved checking for missing values, followed by the identification and replacement of null values.

Additionally, a function was crafted to fill missing zipcodes by leveraging a locality-zipcode mapping. The dtype of the "zipcode" column was then successfully transformed from object to integer, with meticulous handling of various data anomalies, including phrases and alphanumeric entries. The process revealed that a substantial number of rows were converted to zero, prompting considerations for future handling.

Concerning duplicates, it was acknowledged that exact duplicates existed in the dataset, but a decision was made not to remove them, recognizing the potential loss of valuable information within the 151,533 rows. The refined and cleaned dataset, free from inconsistencies, was ultimately saved as a new CSV file for future analyses and error findings.

In summary, the data cleaning process involved addressing missing values, handling diverse data anomalies, converting data types, and making informed decisions regarding duplicates. The resulting dataset is now prepared for further analysis with improved quality and reliability.